### PR TITLE
Support on-chain script decompression and optional pre-compression

### PR DESCRIPTION
**Pull request by lindsgil**
## Description of the change

Updates GenArt721CoreV3, GenArt721CoreV3_Engine, GenArt721CoreV3_Engine_Flex, and GenArt721CoreV3_Explorations contracts to support the (optional) ability for artists to add and update project scripts with pre-compresse scripts. 

In order to leverage adding or updating project scripts with pre-compressed scripts, artists should first call `getProjectScriptCompressed` with their project script and pass the output to `addProjectScriptCompressed` or `updateProjectScriptCompressed`, accordingly.

Linear: https://linear.app/art-blocks/issue/PRO-125/support-on-chain-script-decompression

### CODE REVIEW

File: packages/contracts/contracts/GenArt721CoreV3.sol [hunk 1]
```diff
@@ -1061,6 +1062,28 @@ contract GenArt721CoreV3 is
...
+     * @param _compressedScript Pre-compressed script to be added.
+     * Required to be non-empty, but no further validation is performed.
+     * @dev  A script can be pre-compressed by calling `getProjectScriptCompressed`.
+     */
+    function addProjectScriptCompressed(
+        uint256 _projectId,
+        bytes memory _compressedScript
+    ) external {
+        _onlyUnlocked(_projectId);
+        _onlyArtistOrAdminACL(_projectId, this.addProjectScript.selector);
```
Comment (ryley-o) [inline packages/contracts/contracts/GenArt721CoreV3.sol:1077]: update this to `this.addProjectScriptCompressed.selector`

File: packages/contracts/contracts/GenArt721CoreV3.sol [hunk 2]
```diff
@@ -1479,6 +1527,18 @@ contract GenArt721CoreV3 is
         scriptCount = project.scriptCount;
     }
 
+    /**
+     * @notice Returns the compressed form of a string in bytes using solady LibZip's flz compress algorithm.
+     * @param _script Script to be compressed. Required to be a non-empty string, but no further validaton is performed.
+     * @return bytes compressed bytes
+     */
+    function getProjectScriptCompressed(
```
Comment (ryley-o) [inline packages/contracts/contracts/GenArt721CoreV3.sol:1535]: wdyt about naming this `getCompressed` or `compressScript` or something? I could imagine someone thinking this name is indicating "get the project script for a script that was uploaded as compressed"?

File: packages/contracts/contracts/GenArt721CoreV3.sol [hunk 3]
```diff
@@ -1084,6 +1107,31 @@ contract GenArt721CoreV3 is
...
+     * Required to be non-empty, but no further validation is performed.
+     * @dev  A script can be pre-compressed by calling `getProjectScriptCompressed`.
+     */
+    function updateProjectScriptCompressed(
+        uint256 _projectId,
+        uint256 _scriptId,
+        bytes memory _compressedScript
+    ) external {
+        _onlyUnlocked(_projectId);
+        _onlyArtistOrAdminACL(_projectId, this.updateProjectScript.selector);
```
Comment (ryley-o) [inline packages/contracts/contracts/GenArt721CoreV3.sol:1124]: this needs to be `this.updateProjectScriptCompressed.selector`

File: packages/contracts/contracts/GenArt721CoreV3.sol [hunk 4]
```diff
@@ -1061,6 +1062,28 @@ contract GenArt721CoreV3 is
         emit ProjectUpdated(_projectId, FIELD_PROJECT_SCRIPT);
     }
 
+    /**
+     * @notice Adds a pre-compressed script to project `_projectId`.
```
Comment (ryley-o) [inline packages/contracts/contracts/GenArt721CoreV3.sol:1066]: I think it is worth adding some more details to this function's notice (+ updateProjectScriptCompressed) about what uploading the compressed script is doing. I think some important details to include are:
* a string can be pre-compressed (currently in an @dev comment, I think should just be included in @notice)
* the compressed string is stored in a compressed form on-chain
* no change to read behavior: during reads, the compressed representation is ran through an on-chain decompression algorithm, and the uncompressed string is returned without reliance on any external dependencies

File: packages/contracts/test/core/V3/GenArt721CoreV3_ProjectConfigure.test.ts [hunk 5]
```diff
@@ -1242,6 +1243,270 @@ for (const coreContractName of coreContractsToTest) {
...
+            compressedSkulptuurScript
+          );
+        const script = await config.genArt721Core.projectScriptByIndex(
+          config.projectZero,
+          0
+        );
+        expect(script).to.equal(SKULPTUUR_SCRIPT_APPROX);
+      });
+
+      it("uploads and recalls 23.95 KB script", async function () {
```
Comment (ryley-o) [inline packages/contracts/test/core/V3/GenArt721CoreV3_ProjectConfigure.test.ts:1336]: you might want to skip this on coverage due to it being a memory-heavy operation. was having some issues with memory limits when testing the library itself.

File: packages/contracts/test/core/V3/GenArt721CoreV3_ProjectConfigure.test.ts [hunk 6]
```diff
@@ -1242,6 +1243,270 @@ for (const coreContractName of coreContractsToTest) {
...
+          );
+        const script = await config.genArt721Core.projectScriptByIndex(
+          config.projectZero,
+          0
+        );
+        expect(script).to.equal(CONTRACT_SIZE_LIMIT_SCRIPT);
+      });
+
+      // skip on coverage because contract max sizes are ignored
+      it("fails to upload 50 KB (pre-compressed) script [ @skip-on-coverage ]", async function () {
```
Comment (ryley-o) [inline packages/contracts/test/core/V3/GenArt721CoreV3_ProjectConfigure.test.ts:1358]: I think this is actually 70 kb?

File: packages/contracts/contracts/GenArt721CoreV3.sol [hunk 7]
```diff
@@ -1479,6 +1527,18 @@ contract GenArt721CoreV3 is
         scriptCount = project.scriptCount;
     }
 
+    /**
+     * @notice Returns the compressed form of a string in bytes using solady LibZip's flz compress algorithm.
```
Comment (ryley-o) [inline packages/contracts/contracts/GenArt721CoreV3.sol:1531]: WDYT about noting that the bytes returned from this function are intended to be used as inputs to the add script functions

**Review by ryley-o — COMMENTED** [review]
woohoo! 🎈 Added a few comments and would love to do a final pass after they are resolved 🙏 😃

File: packages/contracts/contracts/GenArt721CoreV3.sol (hunk 1 above)
Comment (lindsgil) [reply packages/contracts/contracts/GenArt721CoreV3.sol]: ah great catch, ty!

File: packages/contracts/contracts/GenArt721CoreV3.sol (hunk 2 above)
Comment (lindsgil) [reply packages/contracts/contracts/GenArt721CoreV3.sol]: Good call! Updated this to `getCompressed`

File: packages/contracts/contracts/GenArt721CoreV3.sol (hunk 4 above)
Comment (lindsgil) [reply packages/contracts/contracts/GenArt721CoreV3.sol]: Good call! Updated the notice tag to include these notes 🙏

File: packages/contracts/test/core/V3/GenArt721CoreV3_ProjectConfigure.test.ts (hunk 5 above)
Comment (lindsgil) [reply packages/contracts/test/core/V3/GenArt721CoreV3_ProjectConfigure.test.ts]: 🙏

File: packages/contracts/test/core/V3/GenArt721CoreV3_ProjectConfigure.test.ts (hunk 6 above)
Comment (lindsgil) [reply packages/contracts/test/core/V3/GenArt721CoreV3_ProjectConfigure.test.ts]: Good catch! Updated 🙏

File: packages/contracts/contracts/GenArt721CoreV3.sol (hunk 7 above)
Comment (lindsgil) [reply packages/contracts/contracts/GenArt721CoreV3.sol]: Good call - updated the notice here to include intended use

**Review by ryley-o — APPROVED** [review]
LGTM! Entering a new era of on-chain project script storage!

File: packages/contracts/contracts/GenArt721CoreV3.sol [hunk 8]
```diff
@@ -1100,6 +1101,32 @@ contract GenArt721CoreV3 is
         emit ProjectUpdated(_projectId, FIELD_PROJECT_SCRIPT);
     }
 
+    /**
+     * @notice Adds a pre-compressed script to project `_projectId`. The script should be compressed using `getCompressed`. This function stores the script in a compressed
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/GenArt721CoreV3.sol]: Ooop, thanks for catching this 🙏  I updated the comments for the add and update functions to adhere to < 80 len

File: packages/contracts/contracts/GenArt721CoreV3.sol [hunk 9]
```diff
@@ -1100,6 +1101,32 @@ contract GenArt721CoreV3 is
...
+    function addProjectScriptCompressed(
+        uint256 _projectId,
+        bytes memory _compressedScript
+    ) external {
+        _onlyUnlocked(_projectId);
+        _onlyArtistOrAdminACL(
+            _projectId,
+            this.addProjectScriptCompressed.selector
+        );
+        require(_compressedScript.length > 0, "Must input non-empty script");
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/GenArt721CoreV3.sol]: 🙏 Awesome, added this! 🙏

**Pull request merged**